Assets/Thirdweb/Core/Scripts/Transaction.cs

```diff
@@ -0,0 +1,254 @@
+        {
+            Input.Data = data;
+            return this;
+        }
+
+        public Transaction SetValue(string ethValue)
+        {
+            Input.Value = BigInteger.Parse(ethValue.ToWei()).ToHexBigInteger();
+            return this;
+        }
+
+        public Transaction SetFrom(string from)
+        {
+            Input.From = from;
+            return this;
+        }
+
+        public Transaction SetGas(string gas)
+        {
+            Input.Gas = BigInteger.Parse(gas).ToHexBigInteger();
+            return this;
+        }
+
+        public Transaction SetTo(string to)
+        {
+            Input.To = to;
+            return this;
+        }
+
+        public Transaction SetType(TransactionType type)
+        {
+            Input.Type = new HexBigInteger((int)type);
+            return this;
+        }
+
+        public Transaction SetGasPrice(string gasPrice)
+        {
+            Input.GasPrice = BigInteger.Parse(gasPrice).ToHexBigInteger();
+            return this;
+        }
+
+        public Transaction SetChainId(string chainId)
+        {
+            Input.ChainId = BigInteger.Parse(chainId).ToHexBigInteger();
+            return this;
+        }
+
+        public Transaction SetNonce(string nonce)
+        {
+            Input.Nonce = BigInteger.Parse(nonce).ToHexBigInteger();
+            return this;
+        }
+
+        public Transaction SetArgs(params object[] args)
+        {
+            var web3 = new Web3(ThirdwebManager.Instance.SDK.session.RPC);
+            var function = web3.Eth.GetContract(contract.abi, contract.address).GetFunction(Input.To);
+            Input.Data = function.GetData(args);
+            return this;
+        }
+
+        public async Task<BigInteger> GetGasPrice()
+        {
+            var web3 = new Web3(ThirdwebManager.Instance.SDK.session.RPC);
+            var gasPrice = await web3.Eth.GasPrice.SendRequestAsync();
+            var maxGasPrice = BigInteger.Parse("300000000000"); // 300 Gwei in Wei
+            var extraTip = gasPrice.Value / 10; // +10%
+            var txGasPrice = gasPrice.Value + extraTip;
+            return txGasPrice > maxGasPrice ? maxGasPrice : txGasPrice;
+        }
+
+        public async Task<BigInteger> EstimateGasLimit()
+        {
+            var gasEstimator = new Web3(ThirdwebManager.Instance.SDK.session.RPC);
+            var gas = await gasEstimator.Eth.Transactions.EstimateGas.SendRequestAsync(Input);
+            var defaultGas = BigInteger.Parse("100000");
+            return gas.Value < defaultGas ? defaultGas : gas;
+        }
+
+        public async Task<GasCosts> EstimateGasCosts()
+        {
+            var gasLimit = await EstimateGasLimit();
+            var gasPrice = await GetGasPrice();
+            var gasCost = gasLimit * gasPrice;
+
+            return new GasCosts { ether = gasPrice.ToString().ToEth(18, false), wei = gasCost };
+        }
+
+        public async Task<Transaction> EstimateAndSetGasAsync()
```

Comment:
when do I use this?

Reply:
If you manually change something like data I guess. I'll also probably use it for my typed transactions later.